### PR TITLE
Add shellcheck to static analysis suite

### PR DESCRIPTION
**Pull request by johnsaigle**
## Brief summary of changes

Building on the work of #5160, this adds `shellcheck` to our static analysis pipeline in the fashion of our PHP linters.

We can iterate on this to fix errors on a file-by-file basis. 

### Testing

Run the install script and make sure it works like it did before.

### CODE REVIEW

**Comment by gdevenyi** [comment]
Rather than checking files present in the list, perhaps instead we should have an exclude list (with a warning of files being excluded). This would allow new shell scripts to be automatically checked by CI, rather than silently skipped.

**Comment by johnsaigle** [comment]
Sure. I'm struggling with the bash but I agree that would be better.

**Comment by gdevenyi** [comment]
I'll think about a way for that!

**Comment by gdevenyi** [comment]
I think I'd do something like this (ref: https://stackoverflow.com/a/15394738/4130016)

```
#!/usr/bin/env bash
set -uo pipefail
shopt -s globstar nullglob
GLOBIGNORE="vendor*:node_modules*"

declare -a excluded=(
    'test/dockerized-integration-tests.sh'
    'test/run-php-linter.sh'
    'test/unittests.sh'
    'test/run-shell-linter.sh'
    'test/dockerized-unit-tests.sh'
    'test/integration.sh'
    'test/run-js-linter.sh'
    'test/wait-for-services.sh'
    'tools/gen-version.sh'
    'tools/install.sh'
    'tools/create-project.sh'
    'tools/deprecated/dump_all_tables_data_only.sh'
    'tools/package_files.sh'
)

files=(**/*.{sh,ksh,bash})

for file in "${files[@]}"; do
  if [[ ! " ${excluded[*]} " =~ ${file} ]]; then
    echo "${file}"
    shellcheck "${file}"
  fi
```

**Comment by gdevenyi** [comment]
Alternative that can trigger the error code:

```sh
#!/usr/bin/env bash
set -euo pipefail
shopt -s globstar nullglob
GLOBIGNORE="vendor*:node_modules*"

#Declare excluded files from check
declare -a excluded=(
    'test/dockerized-integration-tests.sh'
    'test/run-php-linter.sh'
    'test/unittests.sh'
    'test/run-shell-linter.sh'
    'test/dockerized-unit-tests.sh'
    'test/integration.sh'
    'test/run-js-linter.sh'
    'test/wait-for-services.sh'
    'tools/gen-version.sh'
    'tools/install.sh'
    'tools/create-project.sh'
    'tools/deprecated/dump_all_tables_data_only.sh'
    'tools/package_files.sh'
)

#Build array of files
files=(**/*.{sh,ksh,bash})

#Delete array elements present in exclude
for exclude in "${excluded[@]}"; do
  files=( "${files[@]/${exclude}}" )
done

#Cleanup null array entries after delete
for i in "${!files[@]}"; do
  [[ -n "${files[$i]}" ]] || unset "files[$i]"
done

#Run shellcheck on files
echo "${files[@]}"
shellcheck "${files[@]}"

```

**Comment by driusan** [comment]
It needs to error out with an appropriate exit code if any scripts fail or regressions won't be caught by Travis.

**Comment by gdevenyi** [comment]
My second version handles the error code properly.

**Comment by johnsaigle** [comment]
The solution works but I'd like a better option for building list of scripts to scan. In your solution I believe it is also scanning folders like vendor/, node_modules/, and .git/log/ for bash scripts. This slows things down and is generating an error on my system when I run your version. This is why I tried to use the complicated `find` arguments to exclude these directories.

**Comment by gdevenyi** [comment]
``GLOBIGNORE="vendor*:node_modules*"`` handles this. We can add .git to it too.

**Comment by johnsaigle** [comment]
Cool thanks, I can do that part.

File: package.json [hunk 1]
```diff
@@ -33,7 +33,8 @@
   "scripts": {
     "lint:javascript": "./test/run-js-linter.sh",
     "lintfix:javascript": "eslint --fix --ext .js, jsx modules",
-    "lint:php": "./test/run-php-linter.sh",
+    "lint:php": "./test/run-php-linter.php",
```
Comment (johnsaigle) [inline package.json]:
```suggestion
    "lint:php": "./test/run-php-linter.sh",
```

File: test/run-shell-linter.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+set -euo pipefail
+shopt -s globstar nullglob
+GLOBIGNORE="vendor*:node_modules*:.git*"
+
+declare -a excluded=(
+    'test/dockerized-integration-tests.sh'
+    'test/run-php-linter.sh'
+    'test/unittests.sh'
```
Comment (PapillonMcGill) [inline test/run-shell-linter.sh]:
```suggestion
```
Suggest to led by example!

**Comment by gdevenyi** [comment]
> On my VM, dockerized-integration-tests.sh, integration.sh, unittests.sh and unittests.sh reported some error when manually ran which for some reason ignored the excluded list.

I see no such error on manual testing. Can you please provide full input/output showing what happened?

File: test/run-shell-linter.sh (hunk 2 above)
Comment (gdevenyi) [reply test/run-shell-linter.sh]: test/unittests.sh fails shellcheck, are you suggesting it be fixed with this PR?

**Comment by PapillonMcGill** [comment]
@gdevenyi here is the output of `./run-shell-linter.sh`

```
In dockerized-integration-tests.sh line 4:
if [ -v DEBUG ];
        ^-- SC2086: Double quote to prevent globbing and word splitting.


In dockerized-integration-tests.sh line 11:
docker-compose run -T --rm ${CONTAINER} vendor/bin/phpunit --configuration test/phpunit.xml --testsuite LorisCoreIntegrationTests $*
                                                                                                                                  ^-- SC2048: Use "$@" (with quotes) to prevent whitespace problems.
                                                                                                                                  ^-- SC2086: Double quote to prevent globbing and word splitting.


In dockerized-integration-tests.sh line 12:
docker-compose run -T --rm ${CONTAINER} vendor/bin/phpunit --configuration test/phpunit.xml --testsuite LorisModuleIntegrationTests $*
                                                                                                                                    ^-- SC2048: Use "$@" (with quotes) to prevent whitespace problems.
                                                                                                                                    ^-- SC2086: Double quote to prevent globbing and word splitting.


In dockerized-unit-tests.sh line 4:
if [ -v DEBUG ];
        ^-- SC2086: Double quote to prevent globbing and word splitting.


In dockerized-unit-tests.sh line 11:
docker-compose run -T --rm ${CONTAINER} vendor/bin/phpunit --configuration test/phpunit.xml --testsuite LorisUnitTests $*
                                                                                                                       ^-- SC2048: Use "$@" (with quotes) to prevent whitespace problems.
                                                                                                                       ^-- SC2086: Double quote to prevent globbing and word splitting.


In integration.sh line 38:
    l) url="$OPTARG"
       ^-- SC2034: url appears unused. Verify it or export it.


In integration.sh line 52:
export LORIS_DB_CONFIG=$(pwd)/config.xml
       ^-- SC2155: Declare and assign separately to avoid masking return values.


In integration.sh line 67:
mysql -h $host -D $database -u $username -p$password -e "UPDATE Config SET Value='http://localhost:8000' WHERE ConfigID=(SELECT ID FROM ConfigSettings WHERE Name='url')"
         ^-- SC2086: Double quote to prevent globbing and word splitting.
                  ^-- SC2086: Double quote to prevent globbing and word splitting.
                               ^-- SC2086: Double quote to prevent globbing and word splitting.
                                           ^-- SC2086: Double quote to prevent globbing and word splitting.


In integration.sh line 69:
mysql -h $host -D $database -u $username -p$password -e "UPDATE Config SET Value='$(pwd | sed "s#test##")' WHERE ConfigID=(SELECT ID FROM ConfigSettings WHERE Name='base')"
         ^-- SC2086: Double quote to prevent globbing and word splitting.
                  ^-- SC2086: Double quote to prevent globbing and word splitting.
                               ^-- SC2086: Double quote to prevent globbing and word splitting.
                                           ^-- SC2086: Double quote to prevent globbing and word splitting.


In integration.sh line 74:
  echo Running integration test for module: $module;
                                            ^-- SC2086: Double quote to prevent globbing and word splitting.


In integration.sh line 75:
  ../vendor/bin/phpunit --configuration phpunit.xml ../modules/$module/test
                                                               ^-- SC2086: Double quote to prevent globbing and word splitting.


In unittests.sh line 30:
export LORIS_DB_CONFIG=$(pwd)/config.xml
       ^-- SC2155: Declare and assign separately to avoid masking return values.


In unittests.sh line 34:
  echo Running Unit test: $1 in file $2;
                          ^-- SC2086: Double quote to prevent globbing and word splitting.
                                     ^-- SC2086: Double quote to prevent globbing and word splitting.


In unittests.sh line 35:
  ../vendor/bin/phpunit --configuration phpunit.xml $1 $2
                                                    ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                       ^-- SC2086: Double quote to prevent globbing and word splitting.


In unittests.sh line 38:
  echo Running Unit test: $1;
                          ^-- SC2086: Double quote to prevent globbing and word splitting.


In unittests.sh line 39:
  ../vendor/bin/phpunit --configuration phpunit.xml   $1 ./unittests/$1.php
                                                      ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                                     ^-- SC2086: Double quote to prevent globbing and word splitting.


In wait-for-services.sh line 7:
cmd="$@"
    ^-- SC2124: Assigning an array to a string! Assign as array, or use * instead of @ to concatenate.


In wait-for-services.sh line 14:
if [ -v SELENIUM_REQUIRED ]; 
        ^-- SC2086: Double quote to prevent globbing and word splitting.


In wait-for-services.sh line 17:
  until $(curl --output /dev/null --silent --head --fail http://selenium:4444/wd/hub); do
        ^-- SC2091: Remove surrounding $() to avoid executing output.

```

File: test/run-shell-linter.sh (hunk 2 above)
Comment (PapillonMcGill) [reply test/run-shell-linter.sh]: Sorry, wrong line.

File: test/run-shell-linter.sh [hunk 3]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+set -euo pipefail
+shopt -s globstar nullglob
+GLOBIGNORE="vendor*:node_modules*:.git*"
+
+declare -a excluded=(
+    'test/dockerized-integration-tests.sh'
+    'test/run-php-linter.sh'
```
Comment (PapillonMcGill) [inline test/run-shell-linter.sh]:
```suggestion
```
suggest to lead by example!

**Comment by gdevenyi** [comment]
> 09:40:16 [devgab@cicws24:/scratch/Loris] master* ± ./run-shell-linter.sh
> 09:40:18 [devgab@cicws24:/scratch/Loris] master* ±

Are you using the latest version of the code from the PR?

**Comment by johnsaigle** [comment]
@gdevenyi I'm not sure I understand your comment. What was the expected output?

File: test/run-shell-linter.sh (hunk 3 above)
Comment (johnsaigle) [reply test/run-shell-linter.sh]: Done

**Comment by johnsaigle** [comment]
@PapillonMcGill I'm not able to reproduce and Travis seems fine too.

**Comment by PapillonMcGill** [comment]
Yes of course, I used the latest version of the code. Did fetch before checkout.

**Comment by johnsaigle** [comment]
I'm not able to reproduce.

Running `./test/run-shell-linter.sh` gives no errors.

If I delete `tools/gen-version.sh` from the excluded list, I get this error

> In tools/gen-version.sh line 8:
if [ -d '.git/' ] && [ -n $(command -v git) ]; then
                          ^-- SC2070: -n doesn't work with unquoted arguments. Quote or use [[ ]].
                          ^-- SC2046: Quote this to prevent word splitting.

when running `./test/run-shell-linter.sh`, `npm run tests:shell`, and `make checkstatic`, as expected.

**Comment by PapillonMcGill** [comment]
maybe it as to do with version of shell and shellcheck. In my case,

bash 4.3
shellcheck 0.3.7
on Ubuntu 16.04.6

**Comment by johnsaigle** [comment]
bash 4.4.20
Ubuntu 18.04.4
shellcheck 0.4.6

**Comment by PapillonMcGill** [comment]
@johnsaigle Will need to ask IT to upgrade my VM to 18.04.

As we currently `officially support` 16.04, should we at least lint the code so it pass with that version?
I will dismiss my review not to block
@driusan @ridz1208 Should we update our official support to Ubuntu 18.04 (or maybe wait a few more months for 20.04)?

**Comment by johnsaigle** [comment]
I suspect the issue is probably more in the difference of shellcheck or bash versions, not the OS.

**Comment by gdevenyi** [comment]
Just one more datapoint, I get no issues like @johnsaigle on 14.04.5

**Comment by johnsaigle** [comment]
@gdevenyi Which versions of bash and shellcheck do you have? I think this will help us pinpoint what's going on.

**Comment by gdevenyi** [comment]
GNU bash, version 4.4.18(1)-release (x86_64-pc-linux-gnu)
Works for both shellcheck 0.7.0 (local install) and 0.4.4 (ubuntu deb)

**Comment by johnsaigle** [comment]
Strange. @PapillonMcGill maybe the issue is your bash minor version (4.3 vs 4.4)? Or some other configuration issue in your dev environment.

**Comment by johnsaigle** [comment]
@gdevenyi Are you OK if we close your original PR #5160? Your contributions will still be included when this PR is merged.

**Comment by gdevenyi** [comment]
Done

**Comment by johnsaigle** [comment]
@driusan Can you review?

**Comment by johnsaigle** [comment]
@driusan bump

**Pull request merged**